logical/framework/lease.go

```diff
@@ -8,80 +8,100 @@ import (
+	// TTLs
+	now := time.Now().Truncate(time.Second)
+	if startTime.IsZero() {
+		startTime = now
+	} else {
+		startTime = startTime.Truncate(time.Second)
+	}
+
+	// Use the mount's configured max unless the backend specifies
+	// something more restrictive (perhaps from a role configuration
+	// parameter)
+	maxTTL := sysView.MaxLeaseTTL()
+	if backendMaxTTL > 0 && backendMaxTTL < maxTTL {
+		maxTTL = backendMaxTTL
+	}
+	if explicitMaxTTL > 0 && explicitMaxTTL < maxTTL {
+		maxTTL = explicitMaxTTL
+	}
+
+	// Should never happen, but guard anyways
+	if maxTTL <= 0 {
+		return 0, nil, fmt.Errorf("max TTL must be greater than zero")
+	}
 
-		// Use the mount's configured max unless the backend specifies
-		// something more restrictive (perhaps from a role configuration
-		// parameter)
-		max := systemView.MaxLeaseTTL()
-		if backendMax > 0 && backendMax < max {
-			max = backendMax
+	var maxValidTime time.Time
+	switch {
+	case period > 0:
+		// Cap the period value to the sys max_ttl value
+		if period > maxTTL {
+			warnings = append(warnings,
+				fmt.Sprintf("period of %q exceeded the effective max_ttl of %q; period value is capped accordingly", period, maxTTL))
+			period = maxTTL
 		}
+		ttl = period
 
-		// Should never happen, but guard anyways
-		if max < 0 {
-			return nil, fmt.Errorf("max TTL is negative")
+		if explicitMaxTTL > 0 {
+			maxValidTime = startTime.Add(explicitMaxTTL)
+		}
+	default:
+		switch {
+		case increment > 0:
+			ttl = increment
+		case backendTTL > 0:
+			ttl = backendTTL
+		default:
+			ttl = sysView.DefaultLeaseTTL()
 		}
 
 		// We cannot go past this time
-		maxValidTime := leaseOpts.IssueTime.Add(max)
+		maxValidTime = startTime.Add(maxTTL)
+	}
 
-		// Get the current time
-		now := time.Now()
+	if !maxValidTime.IsZero() {
+		// Determine the max valid TTL
+		maxValidTTL := maxValidTime.Sub(now)
 
 		// If we are past the max TTL, we shouldn't be in this function...but
 		// fast path out if we are
-		if maxValidTime.Before(now) {
-			return nil, fmt.Errorf("past the max TTL, cannot renew")
-		}
-
-		// Basic max safety checks have passed, now let's figure out our
-		// increment. We'll use the user-supplied value first, then backend-provided default if possible, or the
-		// mount/system default if not.
-		increment := leaseOpts.Increment
-		if increment <= 0 {
-			if backendIncrement > 0 {
-				increment = backendIncrement
-			} else {
-				increment = systemView.DefaultLeaseTTL()
-			}
+		if maxValidTTL < 0 {
+			return 0, nil, fmt.Errorf("past the max TTL, cannot renew")
 		}
 
-		// We are proposing a time of the current time plus the increment
-		proposedExpiration := now.Add(increment)
-
 		// If the proposed expiration is after the maximum TTL of the lease,
-		// cap the increment to whatever is left
-		if maxValidTime.Before(proposedExpiration) {
-			increment = maxValidTime.Sub(now)
+		// cap the increment to whatever is left, with a small buffer due to
```

Comment:
The comment says that there will be a small buffer, but we aren't adding any buffer.